comp/languagedetection/client/client.go

```diff
@@ -0,0 +1,344 @@
+		freshlyUpdatedPods:           make(map[string]struct{}),
+		periodicalFlushPeriod:        periodicalFlushPeriod,
+	}
+	deps.Lc.Append(fx.Hook{
+		OnStart: cl.start,
+		OnStop:  cl.stop,
+	})
+
+	return util.NewOptional[Component](cl)
+}
+
+// start starts streaming languages to the Cluster-Agent
+func (c *client) start(_ context.Context) error {
+	c.logger.Infof("Starting language detection client")
+	go c.run()
+	return nil
+}
+
+// stop stops the client
+func (c *client) stop(_ context.Context) error {
+	c.cancel()
+	return nil
+}
+
+// run starts processing events and starts streaming
+func (c *client) run() {
+	defer c.logger.Infof("Shutting down language detection client")
+	// workloadmeta can't be initialized in the constructor or provided as a dependency until workloadmeta is refactored as a component
+	if c.store == nil {
+		c.store = workloadmeta.GetGlobalStore() // TODO(components): should be replaced by components
+	}
+
+	eventCh := c.store.Subscribe(
+		subscriber,
+		workloadmeta.NormalPriority,
+		workloadmeta.NewFilter(
+			[]workloadmeta.Kind{
+				workloadmeta.KindKubernetesPod, // Subscribe to pod events to clean up the current batch when a pod is deleted
+				workloadmeta.KindProcess,       // Subscribe to process events to populate the current batch
+			},
+			workloadmeta.SourceAll,
+			workloadmeta.EventTypeAll,
+		),
+	)
+	defer c.store.Unsubscribe(eventCh)
+
+	metricTicker := time.NewTicker(runningMetricPeriod)
+	defer metricTicker.Stop()
+
+	retryProcessWithoutPodTicker := time.NewTicker(c.retryProcessWithoutPodPeriod)
+	defer retryProcessWithoutPodTicker.Stop()
+
+	go c.startStreaming()
+
+	for {
+		select {
+		case eventBundle := <-eventCh:
+			c.processEvent(eventBundle)
+		case <-retryProcessWithoutPodTicker.C:
+			c.retryProcessEventsWithoutPod()
+		case <-c.ctx.Done():
+			err := health.Deregister()
+			if err != nil {
+				c.logger.Warnf("error de-registering health check: %s", err)
+			}
+			return
+		}
+	}
+}
+
+// processEvent processes events from workloadmeta
+func (c *client) processEvent(evBundle workloadmeta.EventBundle) {
+	close(evBundle.Ch)
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	c.logger.Tracef("Processing %d events", len(evBundle.Events))
+	for _, event := range evBundle.Events {
+		switch event.Entity.GetID().Kind {
+		case workloadmeta.KindProcess:
+			c.handleProcessEvent(event, false)
+		case workloadmeta.KindKubernetesPod:
+			c.handlePodEvent(event)
+		}
+
+	}
+}
+
+// startStreaming retrieves the language detection client (= the DCA Client) and periodically sends data to the Cluster-Agent
+func (c *client) startStreaming() {
+	freshUpdateTimer := time.NewTicker(c.newUpdatePeriod)
+	defer freshUpdateTimer.Stop()
+
+	periodicFlushTimer := time.NewTicker(c.periodicalFlushPeriod)
+	defer periodicFlushTimer.Stop()
+
+	health := health.RegisterLiveness("process-language-detection-client-sender")
+
+	ctx, cancel := context.WithCancel(context.Background())
```

Comment:
I think it's not 100% correct in the sense that if `c.ctx` is cancelled, you won't stop the network call until success or timeout (it's the same in the code I liked, to be fair).

You would need something like

Which means that you need something like:
```
ctx, cancel := context.WithCancel(c.ctx)
defer func() {
  cancel()
}

	for {
		select {
		case <-c.ctx.Done():
			err := health.Deregister()
			if err != nil {
				log.Warnf("error de-registering health check: %s", err)
			}
			return
		}

		case healthDeadline := <-health.C:
			cancel()
			ctx, cancel = context.WithDeadline(c.ctx, healthDeadline)
```

The order in the select is important, as we should not try to do anything else even if the parent context is done.

Reply:
👍 Indeed ! I was thinking about that as well but referred to the previous code. It should be fixed now.